habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -1326,3 +1330,232 @@ def obj_next_to(
+            axis=1,
+        )
+        closest_point: np.ndarray = closest_points_each_tri[
+            closest_point_index
+        ]
+        min_dist = float(np.linalg.norm(point - closest_point))
+
+        # Return the minimum distance
+        return min_dist, closest_point
+
+
+def get_obj_receptacle_and_confidence(
+    sim: habitat_sim.Simulator,
+    obj: habitat_sim.physics.ManagedRigidObject,
+    candidate_receptacles: Dict[str, "Receptacle"],
+    support_surface_id: Optional[int] = None,
+    max_dist_to_rec: float = 0.25,
+    island_index: int = -1,
+) -> Tuple[List[str], float, str]:
+    """
+    Heuristic to match a potential placement point with a Receptacle and provide some confidence.
+
+    :param bottom_point: The bottom center point of the object or equivalent (e.g the candidate raycast point for placement)
+    :param obj: The ManagedRigidObject for which to find a Receptacle match.
+    :param support_surface_id: The object_id of the intended support surface (rigid object, articulated link, or stage_id). If not provided, a downward raycast from object center will determine the first hit as the support surface.
+    :param max_dist_to_rec: The threshold point to mesh distance for an object to be matched with a Receptacle.
+    :param candidate_receptacles: Optionally provide a dict (unique_name to Receptacle) of candidate Receptacles for matching. If not provided, uses sim.receptacles.
+    :param island_index: Optionally provide an island_index for identifying navigable floor points. Default is full navmesh.
+    :return: Tuple containing: (1): list of receptacle strings: "floor,region", Receptacle.unique_name, or None (2): a floating point confidence score [0,1] (3): a message string describing the results for use in a UI tooltip
+    """
+    # collect information about failure of the function
+    info_text = ""
+
+    # get the center point of the object projected on the local bounding box size in the gravity direction
+    grav_vector = mn.Vector3(0, -1, 0)
+    dist, center = get_obj_size_along(sim, obj.object_id, grav_vector)
+    obj_bottom_point = center + grav_vector * dist
+
+    # find a support surface if one was not provided
+    if support_surface_id is None:
+        # TODO: This hack is the solution to the Bullet physics raycast margin bug. It should be migrated into every raycast from habitat-sim and then removed here.
+        y_buffer = 0.08
+        raycast_results = sim.cast_ray(
+            habitat_sim.geo.Ray(
+                center + mn.Vector3(0, y_buffer, 0), grav_vector
+            )
+        )
+        if raycast_results.has_hits():
+            for hit in raycast_results.hits:
+                if hit.object_id != obj.object_id:
+                    support_surface_id = hit.object_id
+                    # print(f"first_hit true dist = {hit.ray_distance - y_buffer}")
+                    if hit.ray_distance >= y_buffer:
+                        break
```

Comment:
FYI, this is the fix/hack which impacts the "above" check. Not integrated there yet, but be on the lookout for a habitat-sim PR.
@jacobkrantz @akshararai

Reply:
FYI: https://github.com/facebookresearch/habitat-sim/pull/2466